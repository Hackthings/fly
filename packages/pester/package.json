{
    "name": "@fly/pester",
    "version": "0.0.0",
    "description": "> TODO: description",
    "author": "Michael Dwan <m@dwan.io>",
    "homepage": "",
    "license": "ISC",
    "main": "lib/index.js",
    "directories": {
        "lib": "lib",
        "test": "__tests__"
    },
    "files": [
        "lib/"
    ],
    "scripts": {
        "clean": "rm -rf ./lib ./dist",
        "build": "tsc -b .",
        "rebuild": "yarn clean && yarn build"
    },
    "dependencies": {
        "node-fetch": "2.2.0"
    },
    "devDependencies": {
        "@types/node-fetch": "2.1.2",
        "typescript": "3.0.1"
    }
}