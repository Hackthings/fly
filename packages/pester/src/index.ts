
import { Server, FileAppStore } from "@fly/core"
import axios, { AxiosResponse } from 'axios'
import { URL } from 'url'
import fetch, { Request, Response, RequestInit } from "node-fetch"

export interface EnvironmentOptions {
  testName: string
  testDir: string
  servers: ServerOptions[]
  hostname?: string
}

export interface RequestInit {
  method?: string
}

export class Environment {
  public readonly testName: string
  public readonly testDir: string

  private readonly servers: TestServer[]
  private readonly hostname: string

  constructor(options: EnvironmentOptions) {
    this.testName = options.testName
    this.testDir = options.testDir
    this.servers = new Array<TestServer>()
    this.hostname = options.hostname || "127.0.0.1"

    for (const serverOptions of options.servers) {
      this.servers.push(new TestServer(serverOptions))
    }

    if (this.servers.length == 0) {
      throw new Error("No servers specified")
    }
  }

  public fetch(url: string, init?: RequestInit): Promise<Response> {
    const transformedUrl = this.transformRequestUrl(url)
    console.debug("[fetch]", transformedUrl)
    return fetch(transformedUrl, init)
  }

  public start(): Promise<any> {
    return Promise.all(this.servers.map(s => s.start()))
  }

  public stop(): Promise<any> {
    return Promise.all(this.servers.map(s => s.stop()))
  }

  public get isRunning() {
    return this.servers.every(s => s.isRunning)
  }

  private transformRequestUrl(inputUrl: string) {
    const url = new URL(inputUrl)
    if (url.protocol !== "http:") {
      throw new Error("Protocol must be http")
    }
    for (const server of this.servers) {
      if (url.host === server.name) {
        url.hostname = this.hostname
        url.port = server.port.toString()
        return url.toString()
      }
    }
    throw new Error(`no test server found for host '${url.host}'`)
  }
}

export interface ServerOptions {
  name: string
  path: string
  port?: number
  host?: string
}

class TestServer {
  private server?: Server
  public readonly name: string
  public readonly path: string
  public readonly port: number
  public readonly host: string

  public constructor(options: ServerOptions) {
    this.name = options.name
    this.path = options.path
    this.port = options.port || randomPort()
    this.host = options.host || `${options.name.toLowerCase()}.test`
  }

  public start(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        const appStore = new FileAppStore(this.path, {
          build: false,
          uglify: false,
          env: "test",
          noReleaseReuse: true,
          noWatch: true
        })
        this.server = new Server({ appStore, inspect: false, monitorFrequency: 0 })
        this.server.listen({ port: this.port }, () => {
          resolve()
        })
      } catch (error) {
        reject(error)
      }
    })
  }

  public stop(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.server) {
        try {
          this.server.close(() => {
            resolve()
          })
        } catch (error) {
          reject(error)
        }
      } else {
        resolve()
      }
    })
  }

  public get isRunning(): boolean {
    return this.server && this.server.listening || false
  }
}

function randomPort() {
  return Math.floor((Math.random() * 1000) + 4000)
}
