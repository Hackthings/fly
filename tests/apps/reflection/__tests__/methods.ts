import { Environment } from "@fly/pester"
import axios, { AxiosResponse } from 'axios'
import * as fs from "fs"
import * as path from "path"

function createEnv() {
    return new Environment({
        testName: "helloWorld",
        testDir: __dirname,
        servers: [
            {
                name: "app",
                path: path.resolve(__dirname, "../app.js")
            }
        ]
    })
}

test("handle GET requests", async () => {
    const env = createEnv()
    await env.start()

    const response = await env.fetch("http://app")
    const jsonBody = await response.json()

    expect(jsonBody).toMatchSnapshot()
    // expect(jsonBody).toMatchSnapshot({
    //     headers: {
    //         date: expect.any(String)
    //     }
    // })

    await env.stop()
})

test("handle PUT requests", async () => {
    const env = createEnv()
    await env.start()

    const response = await env.fetch("http://app", { method: "PUT" })
    const jsonBody = await response.json()

    expect(jsonBody).toMatchSnapshot()

    await env.stop()
})

