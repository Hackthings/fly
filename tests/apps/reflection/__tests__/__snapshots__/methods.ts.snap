// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle GET requests 1`] = `
Object {
  "bodySource": "",
  "credentials": "omit",
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "connection": Array [
      "close",
    ],
    "host": "127.0.0.1:4030",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
    "x-request-id": Array [
      "dd9e298a40561ab945917907",
    ],
  },
  "method": "GET",
  "mode": null,
  "referrer": null,
  "remoteAddr": "::ffff:127.0.0.1",
  "stream": null,
  "url": "http://127.0.0.1:4030/",
}
`;

exports[`handle PUT requests 1`] = `
Object {
  "bodySource": "",
  "credentials": "omit",
  "headers": Object {
    "accept": Array [
      "*/*",
    ],
    "accept-encoding": Array [
      "gzip,deflate",
    ],
    "connection": Array [
      "close",
    ],
    "content-length": Array [
      "0",
    ],
    "host": "127.0.0.1:4338",
    "user-agent": Array [
      "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    ],
    "x-request-id": Array [
      "7742803e631a2203874f5c92",
    ],
  },
  "method": "PUT",
  "mode": null,
  "referrer": null,
  "remoteAddr": "::ffff:127.0.0.1",
  "stream": null,
  "url": "http://127.0.0.1:4338/",
}
`;
